Bottom: c484f6e0444262ea36b836c542ebf9df60b84bba
Top:    2c41a458c92e4a1c5ff26174354af8819a676d0f
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-08-30 20:48:23 +0100

Add a hacky name-mangling thingy for loading boards













---

diff --git a/src/create.c b/src/create.c
index c843582..c0c9e46 100644
--- a/src/create.c
+++ b/src/create.c
@@ -37,6 +37,7 @@
 #include <memory.h>
 #include <setjmp.h>
 #include <stdlib.h>
+#include <stdio.h>
 
 #include "global.h"
 
@@ -55,6 +56,8 @@
 #include "undo.h"
 #include "vendor.h"
 
+#include <glib.h>
+
 #ifdef HAVE_LIBDMALLOC
 #include <dmalloc.h>
 #endif
@@ -663,6 +666,57 @@ CreateNewHoleInPolygon (PolygonType *Polygon)
   return Polygon;
 }
 
+static char *refdes_map_file = "refdes.map";
+static GHashTable *refdes_map_table = NULL;
+
+void RefdesMapInit (void)
+{
+  FILE *mapfile;
+  char *from = NULL;
+  char *to = NULL;
+
+  refdes_map_table = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                            g_free, g_free);
+
+  if (refdes_map_file == NULL)
+    return;
+
+  printf ("Refdes map file %s\n", refdes_map_file);
+
+  mapfile = fopen (refdes_map_file, "rb");
+
+  if (mapfile == NULL) {
+    fprintf(stderr, "Could not open [%s]\n", refdes_map_file);
+    return;
+  }
+
+  while (!feof(mapfile) && !ferror(mapfile)) {
+    free (from);
+    free (to);
+
+    from = NULL;
+    to = NULL;
+
+    fscanf (mapfile, "%ms %ms\n", &from, &to);
+
+    if (from == NULL || to == NULL)
+      continue;
+
+    g_hash_table_insert (refdes_map_table, g_strdup (from), g_strdup (to));
+
+    printf ("Stashing rename command: %s to %s\n", from, to);
+  }
+
+  free (to);
+  free (from);
+}
+
+
+char *RefdesMapMap (char *from)
+{
+  return g_hash_table_lookup (refdes_map_table, from);
+}
+
 /* ---------------------------------------------------------------------------
  * creates an new element
  * memory is allocated if needed
@@ -679,6 +733,18 @@ CreateNewElement (DataTypePtr Data, ElementTypePtr Element,
   printf("Entered CreateNewElement.....\n");
 #endif
 
+  /* HACK: Mapping between refdes */
+  if (NameOnPCB != NULL) {
+    char *to;
+    to = RefdesMapMap (NameOnPCB);
+    if (to != NULL) {
+//      free (NameOnPCB);
+      printf ("Mapping name from %s to %s\n", NameOnPCB, to);
+      NameOnPCB = malloc (sizeof (char) * (strlen (to) + 1));
+      strcpy (NameOnPCB, to);
+    }
+  }
+
   if (!Element)
     Element = GetElementMemory (Data);
 
diff --git a/src/create.h b/src/create.h
index ccab272..6c74228 100644
--- a/src/create.h
+++ b/src/create.h
@@ -63,6 +63,7 @@ PolygonTypePtr CreateNewPolygon (LayerTypePtr, FlagType);
 PointTypePtr CreateNewPointInPolygon (PolygonTypePtr,
 				      Coord, Coord);
 PolygonType *CreateNewHoleInPolygon (PolygonType *polygon);
+void RefdesMapInit (void);
 ElementTypePtr CreateNewElement (DataTypePtr, ElementTypePtr,
 				 FontTypePtr, FlagType, char *, char *,
 				 char *, Coord, Coord, BYTE,
diff --git a/src/file.c b/src/file.c
index 1c68abc..9fd2003 100644
--- a/src/file.c
+++ b/src/file.c
@@ -408,6 +408,9 @@ LoadPCB (char *Filename)
   /* mark the default font invalid to know if the file has one */
   newPCB->Font.Valid = false;
 
+  /* Hack, load the mapping file */
+  RefdesMapInit ();
+
   /* new data isn't added to the undo list */
   if (!ParsePCB (PCB, Filename))
     {
